lib/cufft/fft.jl

```diff
@@ -260,138 +247,151 @@ end
-function plan_fft(X::DenseCuArray{T,N}, region) where {T<:cufftComplexes,N}
+function plan_fft(X::StridedCuArray{T,N}, region) where {T<:cufftComplexes,N}
     K = CUFFT_FORWARD
     xtype =  (T == cufftComplex) ? CUFFT_C2C : CUFFT_Z2Z
     inplace = false
 
-    handle, workarea = create_plan(xtype, size(X), region)
+    Y = FakeCuArray{T}(size(X))
+    handle, workarea = create_plan(xtype, X, Y, region)
 
-    cCuFFTPlan{T,K,inplace,N}(handle, workarea, X, size(X), region, xtype)
+    cCuFFTPlan{T,K,inplace,N}(handle, workarea, X, Y, region, xtype)
 end
 
-function plan_bfft(X::DenseCuArray{T,N}, region) where {T<:cufftComplexes,N}
+function plan_bfft(X::StridedCuArray{T,N}, region) where {T<:cufftComplexes,N}
     K = CUFFT_INVERSE
     inplace = false
     xtype =  (T == cufftComplex) ? CUFFT_C2C : CUFFT_Z2Z
 
-    handle, workarea = create_plan(xtype, size(X), region)
+    Y = FakeCuArray{T}(size(X))
+    handle, workarea = create_plan(xtype, X, Y, region)
 
-    cCuFFTPlan{T,K,inplace,N}(handle, workarea, X, size(X), region, xtype)
+    cCuFFTPlan{T,K,inplace,N}(handle, workarea, X, Y, region, xtype)
 end
 
 # out-of-place real-to-complex
-function plan_rfft(X::DenseCuArray{T,N}, region) where {T<:cufftReals,N}
+function plan_rfft(X::StridedCuArray{T,N}, region) where {T<:cufftReals,N}
     K = CUFFT_FORWARD
     inplace = false
     xtype =  (T == cufftReal) ? CUFFT_R2C : CUFFT_D2Z
 
-    handle, workarea = create_plan(xtype, size(X), region)
-
+    halfdim = minimum(region)
     ydims = collect(size(X))
-    ydims[region[1]] = div(ydims[region[1]],2)+1
+    ydims[halfdim] = div(ydims[halfdim],2)+1
+    Y = FakeCuArray{complex(T)}(ydims...)
+    handle, workarea = create_plan(xtype, X, Y, region)
 
-    rCuFFTPlan{T,K,inplace,N}(handle, workarea, X, (ydims...,), region, xtype)
+    rCuFFTPlan{T,K,inplace,N}(handle, workarea, X, Y, region, xtype)
 end
 
-function plan_brfft(X::DenseCuArray{T,N}, d::Integer, region::Any) where {T<:cufftComplexes,N}
+function plan_brfft(X::StridedCuArray{T,N}, d::Integer, region::Any) where {T<:cufftComplexes,N}
     K = CUFFT_INVERSE
     inplace = false
     xtype =  (T == cufftComplex) ? CUFFT_C2R : CUFFT_Z2D
+
+    halfdim = minimum(region)
     ydims = collect(size(X))
-    ydims[region[1]] = d
+    ydims[halfdim] = d
+    Y = FakeCuArray{real(T)}(ydims...)
+
+    # brfft will break input, a copy is performed before execution, 
+    # so the istrides should follow the copy's layout
+    X′ = FakeCuArray{T}(size(X)) 
 
-    handle, workarea = create_plan(xtype, (ydims...,), region)
+    handle, workarea = create_plan(xtype, X′, Y, region)
 
-    rCuFFTPlan{T,K,inplace,N}(handle, workarea, X, (ydims...,), region, xtype)
+    rCuFFTPlan{T,K,inplace,N}(handle, workarea, X′, Y, region, xtype) # istride is useless 
 end
 
-# FIXME: plan_inv methods allocate needlessly (to provide type parameters)
-# Perhaps use FakeArray types to avoid this.
+# plan inv use FakeCuArray to avoid allocations now.
```

Comment:
I'd remove this comment, it's not because we implemented a FIXME that we should note that :-)